s2wav/synthesis.py

```diff
@@ -0,0 +1,103 @@
+import numpy as np
+from s2wav import samples, filters, shapes
+
+# TODO: Switch to S2FFT
+import pyssht as ssht
+import so3
+
+
+def synthesis_transform(
+    f_wav: np.ndarray,
+    f_scal: np.ndarray,
+    L: int,
+    N: int = 1,
+    J_min: int = 0,
+    lam: float = 2.0,
+    spin: int = 0,
+    spin0: int = 0,
+    sampling: str = "mw",
+    kernel: str = "s2dw",
+    reality: bool = False,
+    multiresolution: bool = False,
+) -> np.ndarray:
+    r"""Computes the synthesis directional wavelet transform [1,2].
+
+    Specifically, this transform synthesises the signal :math:`_{s}f(\omega) \in \mathbb{S}^2` by summing the contributions from wavelet and scaling coefficients in harmonic space, see equation 27 from `[2] <https://arxiv.org/pdf/1509.06749.pdf>`_.
+
+    Args:
+        f_wav (np.ndarray): Array of wavelet pixel-space coefficients
+            with shape :math:`[n_{J}, 2N-1, n_{\theta}, n_{\phi}]`.
+
+        f_scal (np.ndarray): Array of scaling pixel-space coefficients
+            with shape :math:`[n_{\theta}, n_{\phi}]`.
+
+        L (int): Harmonic bandlimit.
+
+        N (int, optional): Upper azimuthal band-limit. Defaults to 1.
+
+        J_min (int, optional): Lowest frequency wavelet scale to be used. Defaults to 0.
+
+        lam (float, optional): Wavelet parameter which determines the scale factor between consecutive wavelet scales.
+            Note that :math:`\lambda = 2` indicates dyadic wavelets. Defaults to 2.
+
+        spin (int, optional): Spin (integer) of input signal. Defaults to 0.
+
+        spin0 (int, optional): Spin (integer) of output signal. Defaults to 0.
+
+        sampling (str, optional): Spherical sampling scheme from {"mw","mwss"}. Defaults to "mw".
+
+        kernel (str, optional): The wavelet type from {"s2dw"}. Defaults to "s2dw".
+
+        reality (bool, optional): Whether :math:`f \in \mathbb{R}`, if True exploits
+            conjugate symmetry of harmonic coefficients. Defaults to False.
+
+        multiresolution (bool, optional): Whether to store the scales at :math:`j_{\text{max}}`
+            resolution or its own resolution. Defaults to False.
+
+    Returns:
+        np.ndarray: Signal :math:`f` on the sphere with shape :math:`[n_{\theta}, n_{\phi}]`.
+
+    Notes:
+        [1] B. Leidstedt et. al., "S2LET: A code to perform fast wavelet analysis on the sphere", A&A, vol. 558, p. A128, 2013.
+        [2] J. McEwen et. al., "Directional spin wavelets on the sphere", arXiv preprint arXiv:1509.06749 (2015).
+    """
+    assert f_wav.shape == shapes.f_wav(L, N, J_min, lam, sampling)
```

Comment:
Could consider adding an error message here to explain what went wrong in this assertion.

Reply:
Yeah that would be useful wouldn't it! I'll add some now wherever we make asserts like this.